NilClass's grab bag of baseq3a and other features

ernie opened the pull request:
Sorry, this branch got a little bit out of hand. Feel free to cherry-pick, or request an alternate PR for any given feature.

So far:

- cg_hitSounds
- cg_enemyModel (note that this support needed a tweak to lightall_fp.glsl to avoid partially-transparent player icons in scoreboard, please check my work!)
- cg_enemyColors
- cg_teamModel
- cg_teamColors
- cg_deadBodyDarken
- g_unlagged
- cg_followKiller
- server browser (name filters, a filter field)
- cg_drawSpeed
- server features (autojoin, map rotation, warmup changes)
- improved font rendering
- Include ping ms in lagometer
- High-quality visual defaults
- Fixed Quest controller detection (at least, it wasn't working properly in Virtual Desktop)
- Allow binding of Quest thumbrest touch sensors (default is `+alt`)
- Change default binding for Y to taunt, as ~God~ Team Beef intended.
- Fix transparent player icons on scoreboard in floating HUD
- Render floating HUD at 2x resolution (looks much cleaner)
- Replacement for screen blood spatter. Set cg_damageEffect to 1 to enable. 0 (default) is old behavior.
- Fix for target identification to follow weapon direction (#20)
- Fix for loading screens not rendering
- cg_smoothFollow, an alternative follow mode camera tethered to the one being followed. Orbit camera with primary thumbstick, adjust distance with secondary thumbstick
- improvements to virtual screen (push hud to edges and disable laser pointer)
- Enable follow mode cycling. X button toggles between modes
- vr_hudScale (accessible in UI) to allow pushing HUD to be larger for headsets with a larger sweet spot.
- Fixed first-person follow mode (HUD scaling, disable laser pointer, etc)
- Toggle follow mode in spectate/demo with X
- Recenter on followed player with B
- Fix Con_DrawNotify text positioning/sizing in first person follow and HUD draw mode 2
- Clickable scoreboard in spectator mode to set follow (just like in baseq3a, only supports baseq3, not missionpack)
- Fix an odd "snap" behavior of HUD in single player mode when thumbstick smooth-turning.
- Configurable deadzone for thumbsticks (vr_thumbstickDeadzone) defaulting to 15% (0.15).  I'm liking 0.1, but YMMV.
- Improve thumbstick precision, especially for turning, by dropping the old mouse input emulation and using the thumbstick position like gamepad input does. Allows for very rapid 180s on 100 (default) sens, but still provides slow turning once you get used to having a fuller range of motion.

ernie committed:
Add cg_hitSounds from baseq3a

Working on porting some of my favorite VR-compatible features from
[baseq3a](https://github.com/ec-/baseq3a) to q3vr. First up:
damage-based hit sounds. Based on damage done, the pitch of the hit tone
will change.

Usage:

```
// 0 = no change, 1 = high dmg/low sound, 2 = high dmg/high sound
seta cg_hitSounds 1
```

ernie committed:
Add enemy model support code from baseq3a

From baseq3a client docs:

========================

cg_enemyModel [ pm | modelname ]

    sets model for enemies

    "pm" means use current bright colored enemy model if model is known for mod i.e. there is possible to modify model colors

    if model is unknown (like "alien" or so) then it will be forced to "sarge"

========================

cg_enemyColors [colorstring]

    colorstring is a 3 [or 5] char-length string, where

    1st char - head color, all colors forced to '???' if not set
    2nd char - torso color, forced to white if not set
    3rd char - legs color, forced to white if not set
    4th char - optional, color1 override
    5th char - optional, color2 override

    '?' will be replaced to white in FFA games or corresponding team color (red or blue) in team games

    !!! will work only if cg_enemyModel is set !!!

========================

cg_teamModel  [ pm | modelname ]

    sets model for your teammates

========================

cg_teamColors [colorstring]

    syntax and values is the same as for cg_enemyColors

    colorstring[4] and [5] will ALWAYS override color1 and color2 for spectated client

    but head, torso and legs colors will work corectly only if cg_teamModel is set

========================

cg_deadBodyDarken <0|1>

    turn dead bodies into grey color

    works only if cg_enemyModel is set

    default is 1

ernie committed:
First pass at unlagged implementation

ernie committed:
More enemyModel-related changes

Looks like baseq3a has some specific model behavior for when
following/spectating.

ernie committed:
Port cg_followKiller

Note that while we might assume this does some kind of follow cam after
you are killed, it is really just a way to lazily spectate. In spectator
mode, it will follow the attacker in the first kill that happens, then
track that player's killer, and so on.

ernie committed:
Port server browser updates

Filter field is fiddly, but when you're sorting through a list of >1000
servers, better to have than not. My kingdom for an in-game keyboard.

ernie committed:
Add cg_drawSpeed

Usage:

- cg_drawSpeed 0 - Disabled (default)
- cg_drawSpeed 1 - Display in top-right corner
  counter)
- cg_drawSpeed 2 - Display just under where center crosshair would be

It will hide when displaying scoreboard.

ernie committed:
A bunch of baseq3a server features

Added support for:

- g_rotation
- g_autoJoin
- g_warmup

As with baseq3a, removed/deprecated:

- g_doWarmup (deprecated)
- g_teamAutoJoin (removed)

ernie committed:
Port font changes from baseq3a

Also, fix the drawing of an unnecessary team background image in the
draw code.

ernie committed:
Add ping to lagometer

ernie committed:
Improve visual defaults

Fix build issues.

Further visual improvements

Still more visual improvements

ernie committed:
Stop pulling pak8a.pk3 contents into pakQ3VR.pk3

ernie committed:
Improved Quest button mappings

Turns out that (at least in Virtual Desktop) the input detection order
starting with Index ends up treating Quest controllers as Index ones.
This meant the menu button didn't work properly to bring up the menu,
despite the code to support it being present. Since Index maps trackpad
to menu, this should free us up to use the double-bound menu to taunt,
as ~God~ Team Beef intended.

I also added support for detecting the capacitive touch sensors on the
thumbrests, and defaulted them to `+alt` which should make using the
`_ALT` binding functionality more convenient.

ernie committed:
(Very sloppily) fix the broken railgun crosshair

Need to find the source of the issue requiring this magic 25px offset.

Fix bottom gap on crosshair

ernie committed:
Fix broken alpha on in-world HUD

ernie committed:
Render floating HUD at higher (2x) resolution

We have the extra cycles, and the floating hud at 1280x960 looks way
cleaner

ernie committed:
Maintain full opacity for cg_enemyModel changes

ernie committed:
Add damage taken vignette

Replacement for blood spatter. Set cg_damageIndicator to 1 to enable. 0
(default) is old behavior.

Rename cg_damageIndicator to cg_damageEffect

A bit more specific in terms of what it does.

Add UI option for cg_damageEffect

ernie committed:
Identify targets based on where the weapon points

Fixes #20

ernie committed:
Fix non-rendering loading screens

ernie committed:
Add cg_smoothFollow for alternative demo viewing

Camera is tethered to the followed player. Rotate with primary
thumbstick, zoom in and out with secondary thumbstick forward/back. Not
for those with weak VR legs, but a lot of fun for demo viewing.

Allow follow mode cycling, improve virtual screen

The first person follow mode in virtual screen now has its full screen
size used, instead of rendering the HUD in the middle of the view like
it does when you're in-game.

Default to THIRDPERSON_1 VRFM

Also, stop showing laser pointer when in deathcam

Improve spectator/demo viewing following further

Allow recentering by pressing B button, and when toggling through
followed players. Auto-face followed player in third person views.

ernie committed:
Add vr_hudScale

For headsets with a bigger sweet spot, a larger hud can be more legible,
and also push the edges further out to interfere less with gameplay
awareness.

ernie committed:
Auto-recenter on death/teleport.

ernie committed:
Fix virtual screen hud offset issue

ernie committed:
Further fixes for 1st person following

Should really get around to just making the vr->virtual_screen flag the
one source of truth, maybe? Or maybe not. Anyway, didn't like how the
blitting of the full VR view to a 4:3 aspect ratio squished things, so
did entirely too much work to make that not happen.

ernie committed:
Fix missing demo check for Con_DrawNotify

ernie committed:
Make vr.virtual_screen more intuitive (?)

Of two minds about this commit. Basically, wanted to make it so that
vr.virtual_screen is true any time we are rendering the virtual screen
(including in first-person follow mode), but this of course means adding
vr.first_person_following, because sometimes behavior differs between
these cases.

The thing is, I felt like vr.virtual_screen was a trap, since it didn't
_always_ indicate we would be rendering in a virtual screen.

ernie committed:
Make sure menu code renders in 4:3 correctly

Yeesh, this was a lot of effort for a mode nobody will likely use.

Fix incorrect overlay on map switch

ernie committed:
Bring SS_BLEND0 and later forward

Some maps with stencil shaders seem to use 16 as their sort order. It's
cutting through the HUD, which is annoying.

ernie committed:
Add follow mode scoreboard click for follow

Add Smooth Follow menu option (Comfort)

ernie committed:
Fix cg_hitSounds in single player

ernie committed:
Fix single-player-only HUD shift on thumbstick

ernie committed:
Improve smooth thumbstick feel

Lessen dead zone and provide more granular control of turn speed in
smooth turn mode

Allow vr_thumbstickDeadzone configuration

Defaults to 0.15.

Allow full-speed movement with thumbsticks

Especially when using off-hand movement, the rotation math can leave the
inputs being provided coming up just short of 1.0, which ends up meaning
that a VR player's groundspeed tops out a bit under the 320u/s speed
that a keyboard input might yield. It's a small thing, but figured a fix
was worth implementing.

Additionaly, this fix revealed a missing clamp in CL_FinishMove, which
was causing wrapping on full input (meaning someone running forward
might suddenly, and through no intent of their own, start running
backward)

RippeR37 commented on code/cgame/cg_draw.c:
```diff
@@ -554,17 +554,15 @@ void CG_DrawTeamBackground( int x, int y, int w, int h, float alpha, int team )
-		hcolor[0] = 0;
-		hcolor[1] = 0;
-		hcolor[2] = 0;
+		return;
```
this was added specifically to make UI easier to read in VR, if you want to be able to disable it, add a cvar (or ideally a menu option) to allow for this

RippeR37 commented:
Few questions:

* 9646d86b1154c5b16c9bdaf9ca85643a3e84b2a5 - is this fully compatible with ioquake3 servers?
* 0d48c0537dce7a0160d267f3ea8b14a0d074a3d6 - the change to draw "team background" in non-team game was added specifically to make UI easier to read in VR. I'm fine with adding option to disable it, but removing it completely is not ideal.
* d15564702fd2f1ad8a4db2bf094d91024a9cb0a1 - I agree that currently menu is double-bounded, but for some controllers which don't have touchpad/trackpad people won't have a button to open menu to. I think the code should detect what kind of controllers/headset is used and default to bindings for that one instead, as it seems that VR runtimes do have compatibility mappings between different controllers themselves and just map buttons from one to the other and reply as is fall were supported (or at least multiple). I.e. on SteamVR/Index/Knuckles, it also reports as both Index and Quest controllers are supported.
* 9486caeeae29a3b15c12f8fb5f07fa52f428a057 - does this affect demo playback or normal spectator as well?

I'll do a bit more testing, but overall it looks mostly okay. I think some things will be skipped, but we can merge most as-is and think a bit more on the non-obvious things.

ernie commented:
> > Oh, shoot! Which change specifically? Or are we not sure, just _one_ of them? I never actually looked at the desktop view much, but didn't seem to notice that on VDXR/Quest 3. But that doesn't mean it wasn't there.
> 
> Not sure, my guess is enabling multisampling broke it (as there weren't that many other changes that could result in something like that) but I didn't bisect it. I'd guess [a942012](https://github.com/RippeR37/q3vr/commit/a9420122ac5449b10bad663638233cbfd938cc7d).

Gotcha. Will tweak the cvar when I can test and see if that's it.

ernie committed:
Toggle FFA HUD background tint in UI

ernie committed:
Adjust controller detection based on HMD

ernie commented:
@RippeR37 I am a n00b, so take this with a grain of salt, but I don't think we are actually using multisampling anywhere but on the desktop, anyway? Looks like vr_swapchains.c sets `swapchainCI.sampleCount = 1;`. If so, probably best to just revert those values back to 0?

ernie committed:
Use r_ext_framebuffer_multisample

We were not actually using any multisampling settings before. Also,
added a selector to set MSAA to either off / 2 / 4 / 8.

16x multisampling seems to hang the game on an RTX 5090, so I am pretty
sure we don't want it?

Oops. Sorry, gcc.

ernie commented:
@RippeR37 if you want to give this another look and see if I've addressed your feedback, please do! I can start doing some git surgery tomorrow.

ernie committed:
Correct VR Options menu header

ernie committed:
Allow toggling desktop mirror mode

Can toggle from off / windowed / fullscreen now. In windowed mode, will
vid_restart when changing from one to both eyes, to avoid unnecessary
pillarboxing.

Intentional change, but should find out if Ripper likes this or not: the
old mirroring mode aimed to horizontally fill the scree, but this cut
off vertically, making things like the floating HUD scoreboard and the
console displays not show up properly. If we need to make this a toggle,
we can.

ernie committed:
Crosshair updates

Fix very faint crosshair shaders
While the laser pointer is king, the crosshairs should be visible.

Fix inconsistent crosshair size.

Add crosshair color setting and expose crosshair scale slider.

ernie committed:
Fix misported baseq3a damage counting

Just some out-of-order calculations that would have only messed with
single-player mode or listen servers, really.

ernie committed:
More baseq3a updates.

ernie committed:
cg_damagePlums

Adds a Quake Live-like display of damage numbers in single player or on
supported servers.

ernie committed:
Port missing/buggy baseq3a features

Improved client-side prediction code, fix broken team game player
models/colors.

ernie commented on code/q3_ui/ui_controls3.c:
```diff
@@ -223,14 +223,14 @@ static void Controls3_MenuEvent( void* ptr, int notification ) {
 					trap_Cvar_Set("vr_button_map_B", "+movedown"); // Crouch
 					trap_Cvar_Set("vr_button_map_X", "+button2"); // Use Item
-					trap_Cvar_Set("vr_button_map_Y", "togglemenu"); // Menu
+					trap_Cvar_Set("vr_button_map_Y", "+button3"); // Taunt
```
Hmmm, thought I was catching this [here](https://github.com/RippeR37/q3vr/pull/15/files#diff-60fdd12d8ce5407830ab2cbb4ed2b580eb994da4efd993755d770b8636b4d4efR659-R665), but I assume the second part of the problem you mention will bite me there. I don't have more than one headset to test on, which is making this tough :(

ernie commented on code/vr/vr_swapchains.c:
```diff
@@ -234,7 +278,7 @@ void VR_CreateSwapchain(XrSession session, XrBool32 isColor, int64_t format, con
 	}
 	swapchainCI.format = format;
-	swapchainCI.sampleCount = 1;
+	swapchainCI.sampleCount = sampleCount;
```
Weird. Will double check this. Didn't think it worked without.

ernie committed:
Commit to remove swapchain changes for MSAA.

Note that as best I can tell, this does in fact make MSAA not work. I
left the menu option in for verification.

ernie committed:
Fixes to the desktop mirroring

- Ensure buffer clear (avoid leftover images)
- Restore alt+enter toggle
- blit virtual screen to mirror when virtual screen being displayed in
  headset

ernie committed:
Fix single player spectator offhand movement

Previously, vertical movement in single player was based on HMD
pitch. Now, it follows offhand angles like the multiplayer
implementation does.

ernie committed:
Optimization for cg_damagePlums

ernie commented:
The force push didn't change anything but an updated pak8a.pk3 and the last commit. Just wanted to track the current PR up for baseq3a.

ernie committed:
Small consistency fix for Pickup_PersistantPowerup

(can I just say how bothered I am that the code misspells "persistent"
_everywhere_?)

ernie committed:
Remove duplicated logic in IN_VRJoystick

ernie committed:
Final polish, I hope.

- Fixed mistake on baseq3 crosshair colors
- Added menu options in Team 3 Arena for new settings
- Updated build system to build pakQ3VR.pk3 from assets
- Added pak8a.pk3 and pak3a.pk3 to third_party assets

ernie committed:
Fix a couple of small spectating bugs

- Missing crosshair in first person follow
- Incorrect camera positioning in third-person when spectating in a
  single player match (but why would you, really?)

ernie committed:
Some menu-related fixes/improvements

- Make the in-menu view freeze the camera position/angle
- Fix incorrect color selector on team arena in-game menu

Unrelated, but also in this commit is a fix for single-player
VRFM_THIRDPERSON_2 follow that was causing slight movement in the camera
even when the player isn't moving their head.

The end result: a hopefully less disorienting effect when in a "virtual
theater" that is displaying a screen, so that the screen moves around,
but its contents don't. Well, in a public game the other players keep
doing their thing, but your view will stay consistent.

Works across all follow modes, deathcam, spectator, etc.

ernie committed:
Allow for graceful degradation to vanilla server prediction.

Testing on the ioq3quest version of this code was yielding multiple
pickup sounds, etc. Now, we use the baseq3a prediction code in cases
where we see items with information indicating it's supported, but not
in vanilla servers.

ernie committed:
Fix thumbstick-based weapon selection not disabling smooth turning

ernie committed:
Fix inadvertently-removed debug line

ernie committed:
Tweak player UI for better model display

ernie committed:
Server browser fixes

ernie committed:
Virtual keyboard, and lots of polish.

ernie committed:
Fix recenter follow on smoothfollow

ernie committed:
Port latest ioq3quest work

Virtual Screen Mode:

- Fix squished virtual screen content
- Fix intermission to render in-world when no menu is active, but still use
  screen layer when UI/console is open

HUD Rendering:

- Respect vr_hudDrawStatus=0 to fully disable HUD
- Fix 3D model rendering (player heads, weapon icons) with incorrect alpha by
  adding u_IsBlending uniform and using glBlendFuncSeparate for proper specular
- Fix cl_conXOffset scaling to work correctly across all HUD modes

UI/UX:

- Reduce "following" player name text size and reposition lower on screen
- Add click sound feedback for virtual keyboard key presses

ernie committed:
Fix positioning of team background/scores

ernie committed:
Fix OpenAL error on server start

ernie committed:
Better fix for minor virtual screen squish

ernie committed:
Fix broken single-player intermission

The combination of my changes to VR_useScreenLayer and the position
freeze early return in CG_CalcViewValues I added was conspiring to
override the old code's intention to always render the single-player
intermission with the players on their podiums in-world, becaue
UI_SPPostgameMenu_f sets KEYCATCH_UI, which triggers the virtual screen
now.

ernie committed:
Improvements to click-to-follow

ernie committed:
Improvements to Team Arena server browser

ernie committed:
Improve railgun scope visuals

ernie committed:
Properly rebase timeResidual on prediction upgrade

ernie committed:
Improve console dismissal reliability

ernie committed:
Performance WIP

ernie committed:
Initial fix for "cross-eye" rendering

ernie committed:
Change default binding back for Y

ernie committed:
Remove enhanced baseq3a prediction

The lift to make it truly VR compatible just isn't worth it. Performance
is fine without.

ernie committed:
Fix SP intermission double vision

ernie committed:
Fix and improve podium UI rendering

ernie committed:
Fix regression in frozen virtual screen view

ernie committed:
Fix small nitpick with pickup icon position

It overlapped with team overlays slightly.

ernie committed:
Don't render to overlay buffer in virtual screen

ernie committed:
Improve sharpness of virtual screen

ernie committed:
WIP HUD mode 2 fixes

ernie committed:
Fix console text

ernie committed:
Fix cinematic rendering in team arena

ernie committed:
Fix incorrect flag on scope.tga

ernie committed:
Fix HUD mode 2 offset

ernie committed:
Fix overlap in Team Arena server browser

ernie committed:
Team Arena server browser fixes

ernie committed:
Fix Team Arena cinematics in joinserver/createserver

ernie committed:
Account for ultrawide aspect ratios

ernie committed:
Fix loading screen vertical centering

ernie committed:
Match baseq3a powerup spawn logic

ernie committed:
Reset viewport before drawing virtual screen

Without this, there's a brief period during an intermission where the
screen will drop down to the bottom left.

ernie committed:
Minimal HUD while in weapon zoom

This allows various sounds that are tied to their display (like rewards,
for instance) to play, and otherwise improves game awareness while ADS.
Now, you can hear about how impressive you are without ever zooming out.

ernie committed:
Remove leftover prediction code / fix aspect ratio

Remove baseq3a prediction features that were accidentally left in when we
previously removed the prediction system:

1. Remove painIgnore functionality from cgame:
   - Remove painIgnore field from playerEntity_t in cg_local.h
   - Remove painIgnore checks in CG_PainEvent
   - Remove painIgnore assignments in EV_FALL_MEDIUM and EV_FALL_FAR
   - Remove painTime from EV_FALL_MEDIUM (not present in ioq3)
   - Remove painTime check wrapper from EV_JUMP
   - Keep painTime assignment in EV_FALL_FAR (matches ioq3 behavior)

2. Remove s.time2 pickup prediction assignments from qagame:
   - Remove from FinishSpawningItem in g_items.c
   - Remove from TossClientItems in g_combat.c (weapon and powerup drops)
   - Remove from ClientEvents EV_USE_ITEM1 handler in g_active.c

3. Fix virtual screen vertical squish on non-square framebuffers:
   - Compensate by scaling Y by fov_y/fov_x ratio in the model matrix
   - Should work for all aspect ratios (tall, wide, or square)

ernie committed:
Restore color to names in CG_DrawFollow

Anarki made me do it.

ernie committed:
Fix mp+ pak

ernie committed:
Revert sound optimization.

Turns out, this wasn't the culprit for stutters.

ernie committed:
Improve consistency of podium UI placement

ernie committed:
Fix different blending in deathcam scoreboard

ernie committed:
Fix unable to ready if trigger held on match end

ernie commented:
> * Only part of changes to input handling got in, mainly the remap from `togglemenu` to `taunt` was reverted **for now**. I have some ideas on how to do this better but as-is I wasn't able to open game menu on Index which I think is a big no-no (since there is Q3Q, Q3VR will unfortunately favor PCVR by default, but I hope to have great support for all headsets and runtimes, so ideall in the long run we will have per-headset default controls). I'm already working on it.

Yep, I saw this when I checked the diff. I had intended to revert that default but missed the case when thumbsticks were switched. Do you by chance play left-landed?

> I've fixed the CMake to not require zip/7z. Even though I have 7zip on my Windows machine, it didn't detect it. Instead I've used native CMake functionality to create zip archives so this should improve compatibility.

Yep! This was a great change.

Overall, I was really surprised by how few changes you needed to make and I am very happy we were able to merge. At the moment I am playing around with some stuff that might not play out well in hopes of eventually being able to port this code to Steam Frame native vs android layer, but I am happy to jump on any area you'd like an extra set of hands on and help.
